### PR TITLE
feat: Allow export to netlify using `functions` and `edge-functions` as templates from cli

### PR DESCRIPTION
**Pull request by JayaKrishnaNamburu**
## Description

This PR adds support for templates. 
- Extracted out the defaults like `.vercelignore` and `image-loader`. As, if the users re using some other service providers they come as default and won't help.
- Add prompt for selecting a template at the start of the init process.
- Added template for `netlify-functions` and `netlify-edge-functions`. 

## Code Review

- [ ] hi @istarkov , I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview
  
- [ ] hi @TrySound , I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [x] made a self-review
- [x] added inline comments where things may be not obvious (the "why", not "what")

## Before merging
- [x] tested locally

### CODE REVIEW

File: packages/cli/templates/netlify-edge-functions/remix.config.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+          "nanostores",
+          "@nanostores/react",
+          "@jsep-plugin/assignment",
+        ],
+      };
+
+/** @type {import('@remix-run/dev').AppConfig} */
+module.exports = {
+  ignoredRouteFiles: ["**/.*"],
+  ...baseConfig,
```
Comment (JayaKrishnaNamburu) [inline packages/cli/templates/netlify-edge-functions/remix.config.js:20]: edge workers needs all the dependencies to be bundled along with it.

**Comment by istarkov** [comment]
The only issue saas template must be changed too to load image loader from other file.

**Review by istarkov — APPROVED** [review]

**Comment by istarkov** [comment]
You must check that fixtures are working, especially custom template fixture!!!

**Comment by JayaKrishnaNamburu** [comment]
> The only issue saas template must be changed too to load image loader from other file.

Should i change it in the `saas` repo ivan, or some where else 🤔

File: packages/cli/src/prebuild.ts [hunk 2]
```diff
@@ -126,8 +131,15 @@ const mergeJsonFiles = async (sourcePath: string, destinationPath: string) => {
...
+
+  /*
+    Merge destination onto the source.
+    If the template has a dependency on a package that is already in the project.
+    It will not overwrite if it is the other way around.
+  */
+
   const content = JSON.stringify(
-    merge(JSON.parse(sourceJson), JSON.parse(destinationJson)),
+    merge(JSON.parse(destinationJson), JSON.parse(sourceJson)),
```
Comment (istarkov) [inline packages/cli/src/prebuild.ts:142]: why changed merge?

Comment (istarkov) [reply packages/cli/src/prebuild.ts:142]: It was exactly right!!!!

File: packages/cli/templates/netlify-edge-functions/app/entry.client.tsx [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { RemixBrowser } from "@remix-run/react";
```
Comment (TrySound) [inline packages/cli/templates/netlify-edge-functions/app/entry.client.tsx:1]: Aren't these entry points defaults in remix?

File: packages/cli/templates/netlify-edge-functions/package.json [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{
+  "scripts": {
+    "build": "remix build",
+    "predev": "rimraf ./public/_redirects",
```
Comment (TrySound) [inline packages/cli/templates/netlify-edge-functions/package.json:4]: Why need to delete these?

File: packages/cli/templates/vercel/image-loader.mjs [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import { imageBaseUrl } from "~/constants.mjs";
+
+/**
+ * @type {import("@webstudio-is/image").ImageLoader}
+ */
+
+export const imageLoader = ({ quality, src, width }) => {
```
Comment (TrySound) [inline packages/cli/templates/vercel/image-loader.mjs:7]: Why split into separate module? Might be harder to maintain.

Comment (TrySound) [inline packages/cli/templates/vercel/image-loader.mjs:7]: Why split into separate module? Might be harder to maintain.

File: packages/cli/templates/netlify-edge-functions/remix.config.js (hunk 1 above)
Comment (istarkov) [reply packages/cli/templates/netlify-edge-functions/remix.config.js:20]: Have you tried 
```
  serverBuildPath: "build/index.js",
  serverConditions: ["worker"],
  serverMainFields: ["browser", "module", "main"],
  serverModuleFormat: "esm",
  serverPlatform: "neutral",
  serverDependenciesToBundle: "all",
  serverMinify: true,
```
I used above config to build for worker env

Comment (JayaKrishnaNamburu) [reply packages/cli/templates/netlify-edge-functions/remix.config.js:20]: The `baseConfig` that's coming from the `netlify-edge-adapter` has something for their own. Like they have `deno` in `serverCondition` and then in `serverBuildPath` it's `.netlify-edge-function/script.js` etc. And these are needed only for prod build and production serve according to their template.

Comment (JayaKrishnaNamburu) [reply packages/cli/templates/netlify-edge-functions/remix.config.js:20]: https://github.com/netlify/remix-compute/blob/main/packages/remix-edge-adapter/src/defaultRemixConfig.ts#L4

File: packages/cli/templates/vercel/image-loader.mjs (hunk 5 above)
Comment (JayaKrishnaNamburu) [reply packages/cli/templates/vercel/image-loader.mjs:7]: The `image-loader` is different for deployment targets. The loader for `vercel` is different from netlify and others. So, just using extracted it. So, there can be a default and one that is custom to target 👍 . We can still do it with `constants.mjs` But that moves around a cop of `assetsBase` and other constants everytime so

**Comment by JayaKrishnaNamburu** [comment]
@istarkov updated all fixtures, and added new fixtures for all templates, `custom`, `vercel`, `netlify-functions` and `netlify-edge-functions`.

File: packages/cli/templates/vercel/image-loader.mjs (hunk 5 above)
Comment (TrySound) [reply packages/cli/templates/vercel/image-loader.mjs:7]: It's not like assetsBase take a lot of logic to implement. I'd still have them in one place.

File: fixtures/webstudio-remix-netlify-edge-functions/package.json [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  },
+  "devDependencies": {
+    "@remix-run/serve": "^1.19.2",
+    "@remix-run/dev": "^1.19.2",
+    "@types/react": "^18.2.20",
+    "@types/react-dom": "^18.2.7",
+    "@webstudio-is/http-client": "workspace:*",
+    "eslint": "^8.48.0",
+    "typescript": "5.2.2",
+    "rimraf": "^4.1.4",
```
Comment (TrySound) [inline fixtures/webstudio-remix-netlify-edge-functions/package.json:38]: Node: supports fs.rm with recursive flag
rimraf:
image

Comment (JayaKrishnaNamburu) [reply fixtures/webstudio-remix-netlify-edge-functions/package.json:38]: I get it, but don't we need 1:1 with the original template from netlify. Or we just want to deploy to netlify and make modifications as we need ?

File: packages/cli/templates/vercel/image-loader.mjs (hunk 5 above)
Comment (JayaKrishnaNamburu) [reply packages/cli/templates/vercel/image-loader.mjs:7]: ok, i just followed the suggestion, not to repeat stuff in multiple palaces like `deps` in `package.json` thing. Will change it back then 👍

File: packages/cli/templates/netlify-functions/package.json [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+    "build": "remix build",
+    "predev": "rimraf ./public/_redirects",
+    "dev": "remix dev",
+    "start": "netlify serve",
+    "typecheck": "tsc -b"
+  },
+  "dependencies": {
+    "@netlify/functions": "^1.3.0",
+    "@netlify/remix-adapter": "^1.0.0",
+    "cross-env": "^7.0.3"
```
Comment (TrySound) [inline packages/cli/templates/netlify-functions/package.json:12]: Seem unused

File: fixtures/webstudio-remix-netlify-edge-functions/package.json (hunk 6 above)
Comment (TrySound) [reply fixtures/webstudio-remix-netlify-edge-functions/package.json:38]: Wow, we can do this now
```
node -e "fs.rmSync('./docs', { recursive: true, force: true })"
```

File: packages/cli/templates/netlify-functions/package.json (hunk 7 above)
Comment (JayaKrishnaNamburu) [reply packages/cli/templates/netlify-functions/package.json:12]: yeah, I will clean the netlify templates. Keeping them `1:1` is bringing a lot of un-used stuff too 👍

File: fixtures/webstudio-remix-netlify-edge-functions/app/image-loader.mjs [hunk 8]
```diff
@@ -0,0 +1,8 @@
+import { imageBaseUrl } from "./constants.mjs";
+
+/**
+ * @type {import("@webstudio-is/image").ImageLoader}
```
Comment (TrySound) [inline fixtures/webstudio-remix-netlify-edge-functions/app/image-loader.mjs:4]: No longer used?

File: fixtures/webstudio-remix-netlify-edge-functions/package.json (hunk 6 above)
Comment (TrySound) [reply fixtures/webstudio-remix-netlify-edge-functions/package.json:39]: Still here

File: fixtures/webstudio-remix-netlify-edge-functions/package.json [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+  },
+  "private": true,
+  "sideEffects": false,
+  "devDependencies": {
+    "@remix-run/serve": "^1.19.2",
+    "@remix-run/dev": "^1.19.2",
+    "@types/react": "^18.2.20",
+    "@types/react-dom": "^18.2.7",
+    "@webstudio-is/http-client": "workspace:*",
+    "eslint": "^8.48.0",
```
Comment (TrySound) [inline fixtures/webstudio-remix-netlify-edge-functions/package.json:37]: Eslint again

File: fixtures/webstudio-remix-netlify-edge-functions/package.json [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+    "@webstudio-is/sdk-components-react-remix": "workspace:*",
+    "@webstudio-is/sdk-components-react": "workspace:*",
+    "@webstudio-is/form-handlers": "workspace:*",
+    "@webstudio-is/image": "workspace:*",
+    "@webstudio-is/sdk": "workspace:*",
+    "isbot": "^3.6.8",
+    "react": "^18.2.0",
+    "react-dom": "^18.2.0",
+    "@netlify/remix-adapter": "^1.0.0",
+    "cross-env": "^7.0.3"
```
Comment (TrySound) [inline fixtures/webstudio-remix-netlify-edge-functions/package.json:27]: unused

File: fixtures/webstudio-remix-netlify-edge-functions/app/image-loader.mjs (hunk 8 above)
Comment (JayaKrishnaNamburu) [reply fixtures/webstudio-remix-netlify-edge-functions/app/image-loader.mjs:4]: Deleted them, no longer needed

File: fixtures/webstudio-remix-netlify-edge-functions/package.json (hunk 9 above)
Comment (JayaKrishnaNamburu) [reply fixtures/webstudio-remix-netlify-edge-functions/package.json:37]: We actually discussed about that, but didn't remove from the `templates`. Removing it now 👍

**Review by kof — APPROVED** [review]

File: fixtures/webstudio-remix-vercel/app/constants.mjs [hunk 11]
```diff
@@ -8,18 +8,6 @@ export const imageBaseUrl = "/assets/";
...
-
-  // https://vercel.com/blog/build-your-own-web-framework#automatic-image-optimization
-  return (
-    "/_vercel/image?url=" +
-    encodeURIComponent(imageBaseUrl + src) +
-    "&w=" +
-    width +
-    "&q=" +
-    quality
-  );
```
Comment (TrySound) [inline fixtures/webstudio-remix-vercel/app/constants.mjs:24]: Vercel fixture should have vercel image loader.

**Review by TrySound — APPROVED** [review]

**Pull request merged**